File: BinarySearch/binarySearch.c
#include <stdio.h>
#include <stdlib.h>


int binarySearch(int arr[], int size, int target) {
    int left = 0;
    int right = size - 1;

    while (right >= left) {
        int mid = (left + right) / 2;

        if (arr[mid] == target) {
            return mid;
        }

        if (arr[mid] > target) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    return -1;
}

int main() {
    int arr[10] = {1, 2, 3, 4, 5, 6, 7, 9, 10};
    int size = sizeof(arr) / sizeof(arr[0]); 
    int res = binarySearch(arr, size, 10);

    printf("%d\n", res);

    return 0;
}

File: BinarySearch/BinarySearch.java
package BinarySearch;

import java.util.Arrays;

public class BinarySearch {

    public static void main(String args[]) {
        int[] array = {1, 2, 3, 3, 3, 3, 4, 5, 6};
        int target = 3;
        Arrays.sort(array);
        System.out.println(binarySearch(array, target));
    }

    //TODO: STANDARD BINARY SEARCH TEMPLATE

    public static int binarySearch(int[] array, int target) {
        int left = 0, right = array.length - 1;

        while (left <= right) {
            int mid = (left + right) >> 1;

            if (array[mid] < target) {
                left = mid + 1;
            } else if (array[mid] > target) {
                right = mid - 1;
            } else {
                return mid;
            }
        }
        return -1;
    }
}
